airbyte-integrations/connectors/source-microsoft-sharepoint/source_microsoft_sharepoint/stream_reader.py

```diff
@@ -0,0 +1,197 @@
+        """
+        assert isinstance(value, SourceMicrosoftSharePointSpec)
+        self._config = value
+
+    def list_directories_and_files(self, root_folder, path=None):
+        """Enumerates folders and files starting from a root folder."""
+        drive_items = execute_query_with_retry(root_folder.children.get())
+        found_items = []
+        for item in drive_items:
+            item_path = path + "/" + item.name if path else item.name
+            if item.is_file:
+                found_items.append((item, item_path))
+            else:
+                found_items.extend(self.list_directories_and_files(item, item_path))
+        return found_items
+
+    def get_files_by_drive_name(self, drives, folder_path):
+        """Yields files from the specified drive."""
+        path_levels = [level for level in folder_path.split("/") if level]
+        folder_path = "/".join(path_levels)
+
+        for drive in drives:
+            is_sharepoint = drive.drive_type == "documentLibrary"
+            if is_sharepoint:
+                folder = (
+                    drive.root if folder_path in self.ROOT_PATH else execute_query_with_retry(drive.root.get_by_path(folder_path).get())
+                )
+                yield from self.list_directories_and_files(folder)
+
+    @property
+    @lru_cache(maxsize=None)
+    def drives(self):
+        """
+        Retrieves and caches SharePoint drives, including the user's drive based on authentication type.
+        """
+        drives = execute_query_with_retry(self.one_drive_client.drives.get())
+
+        if self.config.credentials.auth_type == "Client":
+            my_drive = execute_query_with_retry(self.one_drive_client.me.drive.get())
+        else:
+            my_drive = execute_query_with_retry(
+                self.one_drive_client.users.get_by_principal_name(self.config.credentials.user_principal_name).drive.get()
+            )
+
+        drives.add_child(my_drive)
+
+        return drives
+
+    def get_matching_files(self, globs: List[str], prefix: Optional[str], logger: logging.Logger) -> Iterable[RemoteFile]:
+        """
+        Retrieve all files matching the specified glob patterns in SharePoint.
+        """
+        files = self.get_files_by_drive_name(self.drives, self.config.folder_path)
+
+        try:
+            first_file, path = next(files)
+
+            yield from filter_http_urls(
+                self.filter_files_by_globs_and_start_date(
+                    [
+                        MicrosoftSharePointRemoteFile(
+                            uri=path,
+                            download_url=first_file.properties["@microsoft.graph.downloadUrl"],
+                            last_modified=first_file.properties["lastModifiedDateTime"],
+                        )
+                    ],
+                    globs,
+                ),
+                logger,
+            )
+
+        except StopIteration as e:
+            raise AirbyteTracedException(
```

Comment:
It is not clear to me what could cause StopIteration and what it means: is it only because of `get_files_by_drive_name`? Is there a reason why we don't raise this exception as part of `get_files_by_drive_name`? I would see the logic simplified by using a for/else or something like that

Reply:
Fixed